Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesPersistence.cpp

```diff
@@ -0,0 +1,288 @@
+        return reportErrorAndCloseDatabase("prepare delete statement"_s);
+
+    checkedDatabase()->turnOnIncrementalAutoVacuum();
+
+    return true;
+}
+
+void EnhancedSecuritySitesPersistence::deleteSite(const WebCore::RegistrableDomain& site)
+{
+    if (!isDatabaseOpen())
+        return;
+
+    CheckedPtr deleteStatement = cachedStatement(StatementType::DeleteSite).get();
+
+    if (!deleteStatement
+        || deleteStatement->bindText(1, site.string()) != SQLITE_OK
+        || !deleteStatement->executeCommand())
+        reportSQLError("deleteSite"_s, "failed to delete site"_s);
+}
+
+void EnhancedSecuritySitesPersistence::deleteSites(const Vector<WebCore::RegistrableDomain>& sites, CompletionHandler<void()>&& completionHandler)
+{
+    ASSERT(!isMainRunLoop());
+
+    if (!isDatabaseOpen())
+        return completionHandler();
+
+    for (auto& site : sites)
+        deleteSite(site);
+
+    completionHandler();
+}
+
+void EnhancedSecuritySitesPersistence::deleteAllSites(CompletionHandler<void()>&& completionHandler)
+{
+    ASSERT(!isMainRunLoop());
+
+    if (!isDatabaseOpen())
+        return completionHandler();
+
+    auto deleteStatement = checkedDatabase()->prepareStatement(deleteAllSitesSQL);
+    if (!deleteStatement || !deleteStatement->executeCommand())
+        return reportSQLError("deleteAllSites"_s, "delete all sites"_s);
+
+    completionHandler();
+}
+
+HashSet<WebCore::RegistrableDomain> EnhancedSecuritySitesPersistence::fetchEnhancedSecurityOnlyDomains()
+{
+    ASSERT(!isMainRunLoop());
+
+    if (!isDatabaseOpen())
+        return { };
+
+    auto selectStatement = checkedDatabase()->prepareStatement(selectEnhancedSecurityOnlySitesSQL);
+    if (!selectStatement) {
+        reportSQLError("fetchEnhancedSecurityOnlyDomains"_s, "fetch enhanced security only sites"_s);
+        return { };
+    }
+
+    HashSet<WebCore::RegistrableDomain> sites;
+    while (selectStatement->step() == SQLITE_ROW) {
+        auto site = selectStatement->columnText(0);
+        sites.add(WebCore::RegistrableDomain::fromRawString(String { site }));
+    }
+
+    return sites;
+}
+
+HashSet<WebCore::RegistrableDomain> EnhancedSecuritySitesPersistence::fetchAllEnhancedSecuritySites()
+{
+    ASSERT(!isMainRunLoop());
+
+    if (!isDatabaseOpen())
+        return { };
+
+    auto selectStatement = checkedDatabase()->prepareStatement(selectAllSitesSQL);
+    if (!selectStatement) {
+        reportSQLError("fetchAllEnhancedSecuritySites"_s, "fetch all sites"_s);
+        return { };
+    }
+
+    HashSet<WebCore::RegistrableDomain> sites;
+    while (selectStatement->step() == SQLITE_ROW) {
+        auto site = selectStatement->columnText(0);
+        sites.add(WebCore::RegistrableDomain::fromRawString(String { site }));
+    }
+
+    return sites;
+}
+
+void EnhancedSecuritySitesPersistence::trackEnhancedSecurityForDomain(WebCore::RegistrableDomain&& site, EnhancedSecurity reason)
+{
+    ASSERT(!isMainRunLoop());
+
+    if (!isDatabaseOpen())
```

Comment:
Seems like a bad failure case. Maybe add a log, at least? Not sure what else we should do here.